Adding code to generate ascent backend for Catalyst

ayenpure opened the pull request:
-- adding necessary flags `ENABLE_CATALYST` and `CATALYST_DIR` to
   specify if the backend should be created
-- adding the necessary source code into `src/libs/catalyst` directory

ayenpure committed:
Adding code to generate ascent backend for Catalyst

-- adding necessary flags `ENABLE_CATALYST` and `CATALYST_DIR` to
   specify if the backend should be created
-- adding the necessary source code into `src/libs/catalyst` directory

ayenpure commented:
@nicolemarsaglia @cyrush 
Can you please review these changes for building Catalyst's ascent backend?

francois-kitware commented on src/libs/catalyst/AscentCatalyst.cxx:
```diff
@@ -0,0 +1,178 @@
+      std::string content;
+      content.assign((std::istreambuf_iterator<char>(ifs)),
+                     (std::istreambuf_iterator<char>()));
+      conduit::Node node = conduit::Node();
+      node.parse(content, "yaml");
+      node.print();
```
Is it for debug purpose? Should it be removed in the production version?

ayenpure replied:
Yes, it was intended for debug, I've removed it

CharlesGueunet commented on src/libs/catalyst/AscentCatalyst.cxx:
```diff
@@ -0,0 +1,178 @@
+
+    void AddActions(conduit::Node& node)
+    {
+      actions.push_back(node);
```
```suggestion
      this->actions.emplace_back(node);
```

ayenpure replied:
Thanks for the suggestion; I've updated the code

cyrush commented:
This looks good.

Small suggestion (to follow our undocumented CMake standards for TPLs):

Instead of `ENABLE_CATALYST` + `CATALYST_DIR`, we can look for Catalyst if `CATALYST_DIR` is presented.

Then guard further use of Catalyst with `CATALYST_FOUND` (instead of `ENABLE_CATALYST`)


And a Q: I assume `catalyst_implementation` cmake magic hands all the cmake install targets?

ayenpure committed:
Updating CMake flags to comply with the repo

-- Getting rid of `ENABLE_CATALYST` flag and driving the logic using
   `CATALYST_DIR` and `CATALYST_FOUND`

ayenpure committed:
Fixing suggestions from Kitware EU

ayenpure commented:
> This looks good.
> 
> Small suggestion (to follow our undocumented CMake standards for TPLs):
> 
> Instead of `ENABLE_CATALYST` + `CATALYST_DIR`, we can look for Catalyst if `CATALYST_DIR` is presented.
> 
> Then guard further use of Catalyst with `CATALYST_FOUND` (instead of `ENABLE_CATALYST`)
> 
> And a Q: I assume `catalyst_implementation` cmake magic hands all the cmake install targets?

ayenpure commented:
@cyrush, I've updated the branch to remote the `ENABLE_CATALYST` flag.
And yes, the `catalyst_implementation` function does the magic for installing the library at proper targets.
Currently, I have it install into `lib/catalyst` in both `build` and `install` directories.
Let me know if you want me to drop `catalyst` and just install it into `lib`

cyrush commented:
Looks like you were able to reopen, I think we are good.

cyrush commented on src/libs/CMakeLists.txt:
```diff
@@ -59,3 +59,8 @@ endif()
+# Catalyst's ascent implementation
+if(CATALYST_FOUND)
+    add_subdirectory(catalyst)
+endif()
```
Looks good, thanks for these changes!

cyrush commented:
> @cyrush, I've updated the branch to remote the `ENABLE_CATALYST` flag. And yes, the `catalyst_implementation` function does the magic for installing the library at proper targets. Currently, I have it install into `lib/catalyst` in both `build` and `install` directories. Let me know if you want me to drop `catalyst` and just install it into `lib`

Thanks for confirming about the cmake logic. 

 Installing to `lib` would match all the other libs, so if that's possible, let's aim for that.

After that change I will merge.

ayenpure committed:
Changing install dir for libcatalyst-ascent.so

ayenpure commented:
@cyrush, does everything look good to you?

cyrush reviewed (approved):
Looks good, I'll merge pending final CI checks.
Thanks!

cyrush commented:
@ayenpure merged, thanks for adding this!
